### PR TITLE
Denser in-memory representation of `ShardBlobsToDelete`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today each blob to be deleted at the end of a snapshot delete costs us
~80B of heap, and sometimes that can add up to GiBs of temporary heap
usage in total. This commit changes the in-memory representation to use
a compressed stream of pure bytes, which should be more than 4x denser.

Partially mitigates #108278

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Ah, hang on, these thread-local streams are too picky about threading to be used across different threads, even when properly synchronized. I need to work around that.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -1539,31 +1587,88 @@ private record ShardSnapshotMetaDeleteResult(
...
         synchronized void addShardDeleteResult(
             IndexId indexId,
             int shardId,
             ShardGeneration newGeneration,
             Collection<String> blobsToDelete
         ) {
-            shardDeleteResults.add(new ShardSnapshotMetaDeleteResult(indexId, shardId, newGeneration, blobsToDelete));
+            try {
+                shardGenerationsBuilder.put(indexId, shardId, newGeneration);
+                new ShardSnapshotMetaDeleteResult(Objects.requireNonNull(indexId.getId()), shardId, blobsToDelete).writeTo(compressed);
```
Comment (mhl-b) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1618]: As far as I understand blobsToDelete is already occupy memory we want to optimize. Adding in memory compression will temporary increase heap usage by 20%? How do you know that blobsToDelete will be GC-ed after compression?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1618]: There's nothing else retaining it apart from the reference in `ShardBlobsToDelete#shardDeleteResults`.

**Review by mhl-b — APPROVED** [review]

**Review by pxsalehi — APPROVED** [review]
Nice! LGTM.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -1539,31 +1587,88 @@ private record ShardSnapshotMetaDeleteResult(
          *     no further synchronization
          * </p>
          */
-        private final List<ShardSnapshotMetaDeleteResult> shardDeleteResults = new ArrayList<>();
+        private final BytesStreamOutput shardDeleteResults = new ReleasableBytesStreamOutput(bigArrays);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1595]: Sure, see 8e2fe362a3d.

**Pull request merged**